[MTIA Aten Backend][3/n] Migrate mm.out from out-of-tree to in-tree

andyanwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #154393

# Context

See the first PR https://github.com/pytorch/pytorch/pull/153670

# This diff

Migrate mm.out from out-of-tree to in-tree.

We dispatch mm.out to MTIA separately from CPU/CUDA. So this diff adds the file `MTIAOps.cpp` under `ATen/native/mtia` to hold the dispatched functions. In future we can split `MTIAOps.cpp` to categorized ops files.

Differential Revision: [D74743849](https://our.internmc.facebook.com/intern/diff/D74743849/)

cc @egienvalue

andyanwang committed:
[MTIA Aten Backend][3/n] Migrate mm.out from out-of-tree to in-tree

# Context

See the first PR https://github.com/pytorch/pytorch/pull/153670

# This diff

Migrate mm.out from out-of-tree to in-tree.

We dispatch mm.out to MTIA separately from CPU/CUDA. So this diff adds the file `MTIAOps.cpp` under `ATen/native/mtia` to hold the dispatched functions. In future we can split `MTIAOps.cpp` to categorized ops files.

Differential Revision: [D74743849](https://our.internmc.facebook.com/intern/diff/D74743849/)

[ghstack-poisoned]

andyanwang commented:
@pytorchbot label "topic: not user facing"

andyanwang committed:
Update on "[MTIA Aten Backend][3/n] Migrate mm.out from out-of-tree to in-tree"


# Context

See the first PR https://github.com/pytorch/pytorch/pull/153670

# This diff

Migrate mm.out from out-of-tree to in-tree.

We dispatch mm.out to MTIA separately from CPU/CUDA. So this diff adds the file `MTIAOps.cpp` under `ATen/native/mtia` to hold the dispatched functions. In future we can split `MTIAOps.cpp` to categorized ops files.

Differential Revision: [D74743849](https://our.internmc.facebook.com/intern/diff/D74743849/)

cc egienvalue

[ghstack-poisoned]

andyanwang committed:
Update on "[MTIA Aten Backend][3/n] Migrate mm.out from out-of-tree to in-tree"


# Context

See the first PR https://github.com/pytorch/pytorch/pull/153670

# This diff

Migrate mm.out from out-of-tree to in-tree.

We dispatch mm.out to MTIA separately from CPU/CUDA. So this diff adds the file `MTIAOps.cpp` under `ATen/native/mtia` to hold the dispatched functions. In future we can split `MTIAOps.cpp` to categorized ops files.

Differential Revision: [D74743849](https://our.internmc.facebook.com/intern/diff/D74743849/)

cc egienvalue

[ghstack-poisoned]

andyanwang committed:
Update on "[MTIA Aten Backend][3/n] Migrate mm.out from out-of-tree to in-tree"


# Context

See the first PR https://github.com/pytorch/pytorch/pull/153670

# This diff

Migrate mm.out from out-of-tree to in-tree.

We dispatch mm.out to MTIA separately from CPU/CUDA. So this diff adds the file `MTIAOps.cpp` under `ATen/native/mtia` to hold the dispatched functions. In future we can split `MTIAOps.cpp` to categorized ops files.

Differential Revision: [D74743849](https://our.internmc.facebook.com/intern/diff/D74743849/)

cc egienvalue

[ghstack-poisoned]

andyanwang commented on aten/src/ATen/native/mtia/MatrixOps.cpp:
```diff
@@ -0,0 +1,22 @@
+
+DEFINE_DISPATCH(mm_out_mtia_stub);
+
+TORCH_IMPL_FUNC(mm_out_mtia)
+(const at::Tensor& self, const at::Tensor& mat2, const at::Tensor& out) {
+  mm_out_mtia_stub(c10::DeviceType::MTIA, self, mat2, out);
+}
+
+REGISTER_CPU_DISPATCH_NO_OP(
+    mm_out,
+    const at::Tensor&,
+    const at::Tensor& self,
+    const at::Tensor& mat2,
+    const at::Tensor& out);
```
I think we can further combine these two things into one macro. But different ops may need different tweaks. I can try it when add more ops.

andyanwang commented:
The [linux-jammy-py3-clang12-executorch](https://github.com/pytorch/pytorch/actions/runs/15287344092/job/43000851400?pr=154393#logs) CI failure doesn't seem to be related?

malfet commented on aten/src/ATen/native/mtia/MatrixOps.cpp:
```diff
@@ -0,0 +1,22 @@
+#include <ATen/NativeFunctions.h>
```
As this file is not going to be compiled in OSS, can we not export it as well?

andyanwang replied:
n00b question, how can we not export a file under ATen/native? Is there a blocklist of files to be not exported?

malfet commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4157,6 +4157,7 @@
   dispatch:
     CPU: mm_out_cpu
```
Why not extend `mm_out_cpu` and reuse its dispatch for MTI?

nautsimon replied:
Can you elaborate on this, this sounds promising, but we do not fully understand what this might entail

andyanwang replied:
Yeah, @malfet could you help us understand? How can we extend `mm_out_cpu`?

andyanwang commented on aten/src/ATen/native/mtia/MatrixOps.h:
```diff
@@ -0,0 +1,27 @@
+
+namespace at::native {
+
+// This registration is just to make linker happy!
+// Explanation: We use DispatchStub as the bridge between pytorch and mtia's own
+// code. DispatchStub requires DEFAULT(and AVX2 as well if compile with flag
+// HAVE_AVX2_CPU_DEFINITION) to be defined, thus we register for DEFAULT/AVX2
+// here to avoid the "undefined symbol" link error. And these no-op funcitons
+// should never be called anyway, as it's only called when the dispatch key is
+// mtia while the device_type is not mtia.
+#define REGISTER_CPU_DISPATCH_NO_OP(op_name, return_type, ...)             \
+  return_type op_name##_impl_SHOULD_NOT_BE_CALLED(__VA_ARGS__) {           \
+    TORCH_CHECK(false, #op_name ": MTIA op should not be called for CPU"); \
+  }                                                                        \
+  REGISTER_ARCH_DISPATCH(                                                  \
+      op_name##_mtia_stub, DEFAULT, &op_name##_impl_SHOULD_NOT_BE_CALLED)  \
+  REGISTER_AVX2_DISPATCH(                                                  \
+      op_name##_mtia_stub, &op_name##_impl_SHOULD_NOT_BE_CALLED)
```
Basically if not register DEFAULT/AVX2, when compiling this code
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/DispatchStub.h#L228-L231
the linker would complain that DEFAULT/AVX2 is undefined symbol.

andyanwang committed:
Update on "[MTIA Aten Backend][3/n] Migrate mm.out from out-of-tree to in-tree"


# Context

See the first PR https://github.com/pytorch/pytorch/pull/153670

# This diff

Migrate mm.out from out-of-tree to in-tree.

We dispatch mm.out to MTIA separately from CPU/CUDA. So this diff adds the file `MTIAOps.cpp` under `ATen/native/mtia` to hold the dispatched functions. In future we can split `MTIAOps.cpp` to categorized ops files.

Differential Revision: [D74743849](https://our.internmc.facebook.com/intern/diff/D74743849/)

cc egienvalue

[ghstack-poisoned]

andyanwang committed:
Update on "[MTIA Aten Backend][3/n] Migrate mm.out from out-of-tree to in-tree"


# Context

See the first PR https://github.com/pytorch/pytorch/pull/153670

# This diff

Migrate mm.out from out-of-tree to in-tree.

We dispatch mm.out to MTIA separately from CPU/CUDA. So this diff adds the file `MTIAOps.cpp` under `ATen/native/mtia` to hold the dispatched functions. In future we can split `MTIAOps.cpp` to categorized ops files.

Differential Revision: [D74743849](https://our.internmc.facebook.com/intern/diff/D74743849/)

cc egienvalue

[ghstack-poisoned]

albanD reviewed (approved):
Having these specs in native_functions.yaml but not the implmentation is a bit weird. But I guess that's ok for now.

andyanwang committed:
Update on "[MTIA Aten Backend][3/n] Migrate mm.out from out-of-tree to in-tree"


# Context

See the first PR https://github.com/pytorch/pytorch/pull/153670

# This diff

Migrate mm.out from out-of-tree to in-tree.

We dispatch mm.out to MTIA separately from CPU/CUDA. So this diff adds the file `MTIAOps.cpp` under `ATen/native/mtia` to hold the dispatched functions. In future we can split `MTIAOps.cpp` to categorized ops files.

Differential Revision: [D74743849](https://our.internmc.facebook.com/intern/diff/D74743849/)

cc egienvalue

[ghstack-poisoned]

egienvalue reviewed (approved):
LGTM!

andyanwang committed:
Update on "[MTIA Aten Backend][3/n] Migrate mm.out from out-of-tree to in-tree"


# Context

See the first PR https://github.com/pytorch/pytorch/pull/153670

# This diff

Migrate mm.out from out-of-tree to in-tree.

We dispatch mm.out to MTIA separately from CPU/CUDA. So this diff adds the file `MTIAOps.cpp` under `ATen/native/mtia` to hold the dispatched functions. In future we can split `MTIAOps.cpp` to categorized ops files.

Differential Revision: [D74743849](https://our.internmc.facebook.com/intern/diff/D74743849/)

cc egienvalue

[ghstack-poisoned]

andyanwang committed:
Update on "[MTIA Aten Backend][3/n] Migrate mm.out from out-of-tree to in-tree"


# Context

See the first PR https://github.com/pytorch/pytorch/pull/153670

# This diff

Migrate mm.out from out-of-tree to in-tree.

We dispatch mm.out to MTIA separately from CPU/CUDA. So this diff adds the file `MTIAOps.cpp` under `ATen/native/mtia` to hold the dispatched functions. In future we can split `MTIAOps.cpp` to categorized ops files.

Differential Revision: [D74743849](https://our.internmc.facebook.com/intern/diff/D74743849/)

cc egienvalue

[ghstack-poisoned]

andyanwang committed:
Update on "[MTIA Aten Backend][3/n] Migrate mm.out from out-of-tree to in-tree"


# Context

See the first PR https://github.com/pytorch/pytorch/pull/153670

# This diff

Migrate mm.out from out-of-tree to in-tree.

We dispatch mm.out to MTIA separately from CPU/CUDA. So this diff adds the file `MTIAOps.cpp` under `ATen/native/mtia` to hold the dispatched functions. In future we can split `MTIAOps.cpp` to categorized ops files.

Differential Revision: [D74743849](https://our.internmc.facebook.com/intern/diff/D74743849/)

cc egienvalue

[ghstack-poisoned]
